src/FileHelper.php

```diff
@@ -335,21 +342,108 @@ public static function copyDirectory(string $source, string $destination, array
 
             if ($filter === null || $filter->match($from)) {
                 if (is_file($from)) {
-                    if (!$destinationExists) {
-                        self::ensureDirectory($destination, $dirMode);
-                        $destinationExists = true;
-                    }
-                    copy($from, $to);
-                    if ($fileMode !== null) {
-                        chmod($to, $fileMode);
-                    }
+                    self::copyFile($from, $to, $options);
                 } elseif ($recursive) {
                     self::copyDirectory($from, $to, $options);
                 }
             }
         }
 
         closedir($handle);
+
+        self::processCallback($afterCopy, $source, $destination);
+    }
+
+    /**
+     * Copies files with some options.
+     *
+     * - dirMode: integer or null, the permission to be set for newly copied directories. Defaults to null.
+     *   When null - directory will be not created
+     * - fileMode: integer, the permission to be set for newly copied files. Defaults to the current environment
+     *   setting.
+     * - beforeCopy: callback, a PHP callback that is called before copying file. If the callback
+     *   returns false, the copy operation for file will be cancelled. The signature of the
+     *   callback should be: `function ($from, $to)`, where `$from` is the file to be copied from,
+     *   while `$to` is the copy target.
+     * - afterCopy: callback, a PHP callback that is called after file if successfully copied.
+     *   The signature of the callback should be: `function ($from, $to)`, where `$from` is the file
+     *   copied from, while `$to` is the copy target.
+     *
+     * @param string $source The source file
+     * @param string $destination The destination filename
+     * @param array $options
+     *
+     * @psalm-param array{
+     *   dirMode?: int,
+     *   fileMode?: int,
+     *   beforeCopy?: callable,
+     *   afterCopy?: callable,
+     * } $options
+     *
+     * @return bool
+     */
+    public static function copyFile(string $source, string $destination, array $options = []): bool
```

Comment:
```suggestion
    public static function copyFile(string $source, string $destination, array $options = []): void
```

Reply:
But if we need do some action after file copy or not copied. I think it's better like native php`copy` function

Reply:
Maybe create special exception? And if need action - catch this exception.